policy-controller/k8s/api/src/watch.rs

```diff
@@ -18,37 +19,52 @@ where
     W: Stream<Item = Result<Event<T>>> + Send + 'static,
 {
     fn from(watch: W) -> Self {
-        Watch {
-            ready: false,
-            rx: watch.boxed(),
-        }
+        Self::new(watch.boxed())
     }
 }
 
 impl<T> Watch<T> {
-    pub fn ready(&self) -> bool {
-        self.ready
+    pub fn new(rx: Pin<Box<dyn Stream<Item = Result<Event<T>>> + Send + 'static>>) -> Watch<T> {
+        Self {
+            rx,
+            initialized: false,
+            span: tracing::Span::current(),
+        }
+    }
+
+    pub fn instrument(mut self, span: tracing::Span) -> Self {
+        self.span = span;
+        self
+    }
+
+    pub fn is_initialized(&self) -> bool {
+        self.initialized
     }
 
     /// Receive the next event in the stream.
     ///
     /// If the stream fails, log the error and sleep for 1s before polling for a reset event.
     pub async fn recv(&mut self) -> Event<T> {
         loop {
-            match self
+            let ev = self
                 .rx
                 .next()
+                .instrument(self.span.clone())
                 .await
-                .expect("watch stream must not terminate")
-            {
+                .expect(&*format!(
+                    "{} stream must not terminate",
+                    std::any::type_name::<T>()
+                ));
+
+            match ev {
                 Ok(ev) => {
-                    self.ready = true;
+                    self.initialized = true;
                     return ev;
                 }
                 Err(error) => {
-                    self.ready = false;
```

Comment:
👍. So now we'll just keep looping and attempting to restart, but the index loop won't send a readiness update that takes us from ready to not-ready.

Reply:
yeah, the issue is that we won't necessarily get an event when the watch is resumed